### ISSUE TITLE
Poor performance on full GRAB Dataset

### ISSUE THREAD

**Issue by ManuelSenge** [ManipTrans/ManipTrans#69]
Dear authors, 

First, thanks for the amazing work.

I am trying to run ManipTrans on the full GRAB dataset. I followed your instructions on adding a new dataset to the code and created the FullGrabDatasetDexHand based on your demo dataset. The main changes are:

- can load the folder structure of GRAB
- normalizes the scene z_shift to have the object spawn on the simulator table.

See the full Dataset class code at the bottom of the message.

I confirmed that this script spawns the object on the table for 100 GRAB tasks. Also, I visualized the MANO hand using the wrist position and orientation from this dataset. 

With your imitator checkpoint for the shadow hand, I was able to train the model on the airplane fly task:

https://github.com/user-attachments/assets/6cbfdf8f-806a-4d7d-ace6-5302bd51c2cf

However, when trying other tasks, I failed to replicate them. This is the s10 flashlight task:

https://github.com/user-attachments/assets/62ac017d-67c1-402d-91a4-1ed08a6f0e3c

**1. Question: How well does the checkpoint generalize to tasks? Should I be able to run your checkpoint on any GRAB task?** 

When I run your imitator checkpoint without the residual policy in the simulator (the training script as inference), then the resulting wrist positions do not follow the ground truth (I added the mano hand as the ground truth):

https://github.com/user-attachments/assets/38cacc37-f465-4469-a1d0-f9dc2a7b7c9a

To overcome this, I trained my own imitator on the GRAB dataset. I trained it:
- using all 1300+ tasks (purple)
- using only 85 tasks (black)
- using only 3 tasks (blue)

It seems like using fewer tasks helps achieve a better reward and success rate. (I also added the successful episode steps measure, which ignores starting at a later step but only counts actual traversed steps.)

<img width="1441" height="255" alt="Image" src="https://github.com/user-attachments/assets/d425d7db-7aef-4e87-aaf2-93f6894d5058" />

From this, it seems that the more tasks the imitator has, the less reward it gets. However, my feeling is that an imitator trained on fewer tasks does more random wrist movement. I then trained a residual policy on the imitator trained on 3 tasks (also on the others, did not make much of a difference). Here is the result:

https://github.com/user-attachments/assets/fc679750-2a88-432e-ae5f-4f9a474a010c

The model is able (with a lot of random rotations) to follow the wrist movement (mesh as ground truth can be seen in the video). However, the fingers are not really moving or cannot generate the necessary force for grasping. Other tasks are similar, if not worse.

**2. Question: On how many tasks did you train the imitator policy? You mentioned different datasets in the paper.** In [another issue](https://github.com/ManipTrans/ManipTrans/issues/25), you said "6–7 k demonstrations" do you mean tasks?
**3. Question: Did you experience the issue of unwanted wrist rotations during your training?**
**4. Question: Did you experience the issue of little to no finger movement (hindering grasps) yet wrist following for tasks?**

**5. Question: And if so (3./4.), what can one do to compensate for this?**

**6. Question: Does your code require any stiffness and damping tuning for the finger movement? For new Hands but also for the existing one if we see no real finger movement.** With this I mean changing your default parameters: 

```
dexhand_dof_stiffness = torch.tensor(
            [500] * self.dexhand.n_dofs,
            dtype=torch.float,
            device=self.sim_device,
        )
dexhand_dof_damping = torch.tensor(
            [30] * self.dexhand.n_dofs,
            dtype=torch.float,
            device=self.sim_device,
        )
```

I appreciate your answer. Thanks again for your great work.

Manuel

My full GRAB Dataset class:

```
import glob
import os
from functools import lru_cache
from types import SimpleNamespace

import numpy as np
import torch
import trimesh
from manotorch.manolayer import ManoLayer
from pytorch3d.structures import Meshes

from .base import ManipData
from .decorators import register_manipdata
from main.dataset.transform import aa_to_rotmat


class _FullGrabDexHandBase(ManipData):
    mano_side = "right"
    hand_key = "rhand"

    def __init__(
        self,
        *,
        data_dir: str = "data/GRAB",
        split: str = "all",
        skip: int = 2, # temporal downsampling factor (1= 120Hz, 2=60Hz, 4=30Hz)
        device="cuda:0",
        mujoco2gym_transf=None,
        max_seq_len=int(1e10),
        dexhand=None,
        support_clearance: float = 0.0000002,
        support_window: int = 10,
        **kwargs,
    ):
        super().__init__(
            data_dir=data_dir,
            split=split,
            skip=skip,
            device=device,
            mujoco2gym_transf=mujoco2gym_transf,
            max_seq_len=max_seq_len,
            dexhand=dexhand,
            **kwargs,
        )
        self.manolayer = ManoLayer(
            rot_mode="axisang",
            side=self.mano_side,
            center_idx=None,
            mano_assets_root="data/mano_v1_2",
            use_pca=False,
            flat_hand_mean=True,
        ).to(device)

        # select task based on index: total 9 sec --> now 3 sec
        self.cut_beginning_frames = 0 #60 * 3 #int(120*3) #120 * 2 # sec
        self.cut_end_frames = None#-60 * 3  #must be negative120 * 1 # sec before: 1
        # 1270 1330

        self.grab_root = os.path.join(self.data_dir, "grab")
        self.data_pathes = sorted(glob.glob(os.path.join(self.grab_root, "s*", "*.npz")))

        self.device = device
        self.support_clearance = support_clearance
        self.support_window = support_window

        transf_offset = np.eye(4)
        transf_offset[:3, :3] = aa_to_rotmat(np.array([-np.pi / 2, 0, 0])) @ aa_to_rotmat(np.array([0, 0, np.pi / 2]))
        transf_offset[:3, 3] = np.array([0.0, 0.0, 0.0]) #np.array([0.0, 0.018, 0.0])

        self.transf_offset = torch.tensor(transf_offset, dtype=torch.float32, device=mujoco2gym_transf.device)

        self.mujoco2gym_transf = mujoco2gym_transf @ self.transf_offset
        self.max_seq_len = max_seq_len

        mesh_paths = []
        for ext in ("*.stl", "*.obj", "*.ply"):
            mesh_paths.extend(glob.glob(os.path.join(self.data_dir,'tools', 'object_meshes', 'contact_meshes', ext)))
        # find mesh file by name
        self.mesh_map = {os.path.splitext(os.path.basename(path))[0]: path for path in sorted(mesh_paths)}
    

        self.urdf_dir = os.path.join(self.data_dir, "urdf")
        self.urdf_template_path = "assets/obj_urdf_example.urdf"

    @staticmethod
    def _load_mesh_as_trimesh(mesh_path):
        mesh = trimesh.load(mesh_path, process=False)
        if isinstance(mesh, trimesh.Scene):
            mesh = mesh.dump(concatenate=True)
        if not isinstance(mesh, trimesh.Trimesh):
            raise TypeError(f"Expected trimesh.Trimesh, got {type(mesh)} for {mesh_path}")
        return mesh

    def _load_table_data(self, seq_data, length):
        if "table" not in seq_data:
            return None, None

        table_block = seq_data["table"].item()
        table_mesh_path = table_block.get("table_mesh")
        if table_mesh_path is None:
            return None, None

        table_mesh = self._load_mesh_as_trimesh(os.path.join(self.data_dir, table_mesh_path))
        table_params = table_block["params"]

        table_pose = np.eye(4, dtype=np.float32)[None].repeat(length, axis=0)
        table_pose[:, :3, :3] = aa_to_rotmat(table_params["global_orient"]).astype(np.float32)[self.cut_beginning_frames:self.cut_end_frames: self.skip]
        table_pose[:, :3, 3] = table_params["transl"].astype(np.float32)[self.cut_beginning_frames:self.cut_end_frames: self.skip]
        return table_mesh, table_pose

    def _compute_table_alignment_shift(self, seq_data, table_mesh, table_pose, obj_pose, obj_mesh):
        if table_mesh is None or table_pose is None:
            return 0.0

        table_vertices = np.asarray(table_mesh.vertices, dtype=np.float32)
        obj_vertices = np.asarray(obj_mesh.vertices, dtype=np.float32)
        frame_id = 0
        table_world = (table_pose[frame_id, :3, :3] @ table_vertices.T).T + table_pose[frame_id, :3, 3]
        obj_world = (obj_pose[frame_id, :3, :3] @ obj_vertices.T).T + obj_pose[frame_id, :3, 3]

        raw_table_top_z = float(table_world[:, 2].max())
        raw_obj_bottom_z = float(obj_world[:, 2].min())
        support_gap = raw_obj_bottom_z - raw_table_top_z
        lift_z = max(0.0, self.support_clearance - support_gap)
        scene_shift_z = -raw_table_top_z + lift_z

        return scene_shift_z

    def _build_scene_objects(self, seq_data, scene_shift_z):
        # GRAB .npz files currently expose the manipulated object and the table only.
        # The simulator task already instantiates its own support table, so we use
        # the recorded table only for scene normalization and avoid duplicating it
        # as another collidable actor here. Keep the structure in place so future
        # scene elements can be added and receive the same scene-level shift.
        scene_objs = []

        extra_scene_elements = seq_data.get("scene_elements", None)
        if extra_scene_elements is None:
            return scene_objs

        for scene_element in extra_scene_elements:
            pose = np.array(scene_element["pose"], dtype=np.float32, copy=True)
            pose[:3, 3] += np.array([0.0, 0.0, scene_shift_z], dtype=np.float32)
            scene_objs.append(
                {
                    "obj": SimpleNamespace(
                        type=scene_element["type"],
                        size=np.asarray(scene_element["size"], dtype=np.float32),
                    ),
                    "pose": pose,
                }
            )
        return scene_objs

    def _ensure_obj_urdf(self, mesh_path):
        # creates the urdf file if not exist from mesh and returns the urdf path
        #mesh_dir = os.path.dirname(mesh_path)# 'data/GRAB/stl'
        mesh_file = os.path.basename(mesh_path)# 'airplane.stl'
        mesh_stem = os.path.splitext(mesh_file)[0] # 'airplane'

        urdf_path = os.path.join(self.urdf_dir, f"{mesh_stem}.urdf")
        if os.path.exists(urdf_path):
            return urdf_path

        with open(self.urdf_template_path, "r") as f:
            urdf_text = f.read()

        urdf_text = urdf_text.replace("102_obj.obj", mesh_path)

        with open(urdf_path, "w") as f:
            f.write(urdf_text)

        return urdf_path

    def __len__(self):
        return len(self.data_pathes)
    
    @lru_cache(maxsize=None)
    def __getitem__(self, index):
        # remove indicator for index
        if isinstance(index, str):
            index = int(index.replace('fg', ''))
        #index = 108

        seq_path = self.data_pathes[index]
        seq_data = np.load(self.data_pathes[index], allow_pickle=True)

        mesh_path = self.mesh_map[seq_data["obj_name"].item()]
        obj_mesh = self._load_mesh_as_trimesh(mesh_path)

        hand_params = seq_data[self.hand_key].item()["params"] # hand trajectory
        obj_params = seq_data["object"].item()["params"] # object trajectory

        length = len(obj_params["global_orient"][self.cut_beginning_frames:self.cut_end_frames: self.skip])
        print(length, 'the new length')
        obj_pose = np.eye(4, dtype=np.float32)[None].repeat(length, axis=0)
        obj_pose[:, :3, :3] = aa_to_rotmat(obj_params["global_orient"][self.cut_beginning_frames:self.cut_end_frames: self.skip]).astype(np.float32).transpose(0, 2, 1)
        obj_pose[:, :3, 3] = obj_params["transl"][self.cut_beginning_frames:self.cut_end_frames: self.skip].astype(np.float32)
        table_mesh, table_pose = self._load_table_data(seq_data, length)

        hand_pose = torch.tensor(np.concatenate([hand_params["global_orient"], hand_params["fullpose"]], 
                                                axis=1)[self.cut_beginning_frames:self.cut_end_frames: self.skip], dtype=torch.float32, device=self.device)
        hand_transl = torch.tensor(hand_params["transl"][self.cut_beginning_frames:self.cut_end_frames: self.skip], dtype=torch.float32, device=self.device)

        mano_out = self.manolayer(hand_pose)
        mano_out_verts = mano_out.verts + hand_transl[:, None, :] # account for world frame
        mano_out_joints = torch.matmul(self.manolayer.th_J_regressor, mano_out_verts)

        wrist_pos = mano_out_joints[:, 0].detach()
        middle_pos = mano_out_joints[:, 4].detach()
        wrist_pos = wrist_pos - (middle_pos - wrist_pos) * 0.25

        mano_joints = {
            "index_proximal": mano_out_joints[:, 1].detach(),
            "index_intermediate": mano_out_joints[:, 2].detach(),
            "index_distal": mano_out_joints[:, 3].detach(),
            "index_tip": mano_out_verts[:, 353].detach(),
            "middle_proximal": mano_out_joints[:, 4].detach(),
            "middle_intermediate": mano_out_joints[:, 5].detach(),
            "middle_distal": mano_out_joints[:, 6].detach(),
            "middle_tip": mano_out_verts[:, 467].detach(),
            "pinky_proximal": mano_out_joints[:, 7].detach(),
            "pinky_intermediate": mano_out_joints[:, 8].detach(),
            "pinky_distal": mano_out_joints[:, 9].detach(),
            "pinky_tip": mano_out_verts[:, 695].detach(),
            "ring_proximal": mano_out_joints[:, 10].detach(),
            "ring_intermediate": mano_out_joints[:, 11].detach(),
            "ring_distal": mano_out_joints[:, 12].detach(),
            "ring_tip": mano_out_verts[:, 576].detach(),
            "thumb_proximal": mano_out_joints[:, 13].detach(),
            "thumb_intermediate": mano_out_joints[:, 14].detach(),
            "thumb_distal": mano_out_joints[:, 15].detach(),
            "thumb_tip": mano_out_verts[:, 766].detach(),
        }
        inspire_rot_offset = self.dexhand.relative_rotation[None] # adds the time dim
        wrist_rot = torch.tensor(
            aa_to_rotmat(hand_params["global_orient"][self.cut_beginning_frames:self.cut_end_frames: self.skip]),
            dtype=torch.float32,
            device=self.device,) @ torch.tensor(
            np.repeat(inspire_rot_offset, length, axis=0),
            dtype=torch.float32,
            device=self.device,
        )
        wrist_pos += torch.tensor(self.dexhand.relative_translation, device=self.device)

        scene_shift_z = self._compute_table_alignment_shift(seq_data, table_mesh, table_pose, obj_pose, obj_mesh)
        scene_shift = torch.tensor([0.0, 0.0, scene_shift_z], dtype=torch.float32, device=self.device)

        obj_pose = torch.tensor(obj_pose, dtype=torch.float32, device=self.device)
        obj_pose[:, :3, 3] += scene_shift
        wrist_pos += scene_shift
        for joint_name in mano_joints.keys():
            mano_joints[joint_name] += scene_shift
        scene_objs = self._build_scene_objects(seq_data, scene_shift_z)
        
        mesh = Meshes(
            verts=torch.from_numpy(obj_mesh.vertices[None, ...].astype(np.float32)),
            faces=torch.from_numpy(obj_mesh.faces[None, ...].astype(np.int64)),
        )
        rs_verts_obj = self.random_sampling_pc(mesh)

        seq_rel_path = os.path.relpath(seq_path, self.grab_root)
        mano_out_verts = mano_out_verts + scene_shift[None, None, :]

        data = {
            "data_path": seq_path,
            "obj_id": seq_data["obj_name"].item(),
            "obj_verts": rs_verts_obj,
            "obj_urdf_path": self._ensure_obj_urdf(mesh_path),
            "obj_trajectory": obj_pose,
            "scene_objs": scene_objs,
            "wrist_pos": wrist_pos,
            "wrist_rot": wrist_rot,
            "mano_joints": mano_joints,
        }        

        self.process_data(data, index, rs_verts_obj)

        opt_path = os.path.join(
            "data",
            "retargeting",
            "fullgrab",
            f"mano2{str(self.dexhand)}",
            os.path.splitext(seq_rel_path)[0] + ".pkl",
        )
        self.load_retargeted_data(data, opt_path)

        return data


@register_manipdata("fullgrab_rh")
class FullGrabDatasetDexHandRH(_FullGrabDexHandBase):
    mano_side = "right"
    hand_key = "rhand"


@register_manipdata("fullgrab_lh")
class FullGrabDatasetDexHandLH(_FullGrabDexHandBase):
    mano_side = "left"
    hand_key = "lhand"

```

**Comment by KailinLi** [comment]
Hi @ManuelSenge, thanks for your interest in ManipTrans.

First, you may refer to issue [#22](https://github.com/ManipTrans/ManipTrans/issues/22#issuecomment-2830108392)￼, which may provide some useful experience regarding GRAB.

About your questions:

1. One major difference between GRAB and our other datasets is that GRAB often starts from a human arm in a stretched / T-pose-like posture. This means the hand needs to move a relatively long distance at a high speed before reaching the object. This is somewhat out-of-distribution for our setting, since such highly dynamic cases are relatively rare in our data. Therefore, tracking failures can be observed especially during the approaching stage.
    Regarding your observation that fewer tasks may lead to better reward and success rate, this is indeed a trade-off depending on model capacity and action complexity. It is also a very interesting research question.
2. For details about the data, please refer to [#44](https://github.com/ManipTrans/ManipTrans/issues/44#issuecomment-3213354801)￼ and [#53](https://github.com/ManipTrans/ManipTrans/issues/53#issuecomment-3538520737)￼.
3. Yes, we have observed unwanted wrist rotations in some cases. This usually happens when the tracked motion sequence is too fast and becomes out-of-distribution. A possible future direction is to modify the control mode; for example, the simplest option could be mounting the hand on a robotic arm.

4–5. We did not observe the issue of fingers barely moving. My guess is that this may be caused by too many wrist-tracking failures, which prevent the policy from learning the detailed finger motions properly.

6. We use the default stiffness and damping settings in our experiments, and they generally work well. However, you can also tune them for different dexterous hands. In principle, this may lead to better results, especially if you consider real-world deployment later.

Thanks again for your interest and detailed experiments.
